### PR TITLE
Change name combinatorics to topological analysis

### LINKED ISSUE #1951: Documentation restructuring
We want to restructure the documentation so that it follows the setup for msprime, roughly. That is

- A separate part for reference documentation (include Python and C APIs, and CLI) called "Interfaces"
- A part that includes thematically structured documentation that describes features with the help of examples (following the structure, e.g. of https://tskit.dev/msprime/docs/stable/ancestry.html
- Something corresponding to the current miscellaneous section

Essentially, we spent a great deal of effort on the msprime documentation structure so we might as well try and bring as much of that over as we can.

I guess we can close this issue when:

- We've updated the high-level structure as described above
- Created issues to track any remaining things that have cropped up as needing to be done during the process.

### PR DESCRIPTION
**Pull request by hyanwong**
And add example of traversal. This will somewhat replace the [current section in the tutorial](https://tskit.dev/tutorials/analysing_trees.html#tree-traversals), but we decided it would be sensible to have a basic introduction in the docs.

Fixes #1951

### CODE REVIEW

File: docs/topological-analysis.md [hunk 1]
```diff
@@ -14,9 +14,117 @@ kernelspec:
...
+import numpy as np
+av_arity = np.mean([tree.num_children(u) for u in tree.nodes() if not tree.is_leaf(u)])
+print(f"Average arity of internal nodes: {av_arity}")
+```
+
+:::{note}
+In `tskit`, a tree can have multiple {ref}`sec_data_model_tree_roots`, each with a
+descendant topology.  However, for some algorithms, instead of traversing through
+the descendants of each root of a tree in turn, it can be helpful to start at
+single node and traverse downwards through the entire genealogy. The
```
Comment (jeromekelleher) [inline docs/topological-analysis.md:76]:
```suggestion
a single node and traverse downwards through the entire genealogy. The
```

File: docs/topological-analysis.md [hunk 2]
```diff
@@ -14,9 +14,117 @@ kernelspec:
...
+{ref}`virtual root<sec_data_model_tree_virtual_root>` is provided for this purpose.
+:::
+
+#### Array methods
+
+The {meth}`Tree.nodes` iterator provides a convenient way of looping over descendant
+node IDs, but it can be more efficient to deal with all the IDs all at once, as a
```
Comment (jeromekelleher) [inline docs/topological-analysis.md:83]:
```suggestion
node IDs, but it can be more efficient to deal with all the IDs at once, as a
```

**Review by jeromekelleher — APPROVED** [review]
Looks great, thanks. Minor typos

**Comment by hyanwong** [comment]
> Looks great, thanks. Minor typos

Well spotted, and corrected now. Thanks for reading through: much appreciated.

**Pull request merged**